### PR TITLE
Feat: support filter argument for nerdctl volume command

### LINKED ISSUE #1330: Support filter in `nerdctl volume ls` command
### What is the problem you're trying to solve

Just following #54, to support the filter argument in `nerdctl volume ls` command 

- [ ] filtering by label
- [ ] filtering by name

### Describe the solution you'd like

Support filter in `nerdctl volume ls` command

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yuchanns**
Fix #1330 

Signed-off-by: Hanchin Hsieh <me@yuchanns.xyz>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1079,8 +1079,12 @@ Flags:
...
   - :nerd_face: `--format=json`: Alias of `--format='{{json .}}'`
 - :nerd_face: `--size`: Display the disk usage of volumes.
-
-Unimplemented `docker volume ls` flags: `--filter`
+- :whale: `-f, --filter`: Filter volumes based on given conditions.
+    - :whale: `--filter label=<key>=<value>`: Matches volumes by label on both
+      `key` and `value`. If `value` is left empty, matches all volumes with `key`
+      regardless of its value
+    - :whale: `--filter name=<value>`: Matches all volumes with a name containing
+      the `value` string
```
Comment (AkihiroSuda) [inline README.md:1087]: Could you document unimplemented filter such as `dangling`?
https://docs.docker.com/engine/reference/commandline/volume_ls/

See also the document of `nerdctl images --filter`

File: cmd/nerdctl/volume_ls.go [hunk 2]
```diff
@@ -154,3 +164,54 @@ func getVolumes(cmd *cobra.Command) (map[string]native.Volume, error) {
 	}
 	return volStore.List(volumeSize)
 }
+
+func getFilters(filters []string) ([]map[string]string, []string, bool) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_ls.go:168]: Perhaps
```suggestion
func getVolumeFilters(filters []string) ([]map[string]string, []string, bool) {
```

File: cmd/nerdctl/volume_ls.go [hunk 3]
```diff
@@ -154,3 +164,54 @@ func getVolumes(cmd *cobra.Command) (map[string]native.Volume, error) {
...
+				filterLabel := map[string]string{subs[1]: v}
+				labelFilters = append(labelFilters, filterLabel)
+				continue
+			}
+		}
+	}
+	return labelFilters, nameFilters, isFilter
+}
+
+func isVolumeMatches(vol native.Volume, filterLabels []map[string]string, filterNames []string) bool {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_ls.go:192]:
```suggestion
func volumeMatchesFilter(vol native.Volume, filterLabels []map[string]string, filterNames []string) bool {
```

File: cmd/nerdctl/volume_ls_test.go [hunk 4]
```diff
@@ -81,3 +81,276 @@ func TestVolumeLs(t *testing.T) {
...
+		}
+
+		var numMatches = 0
+		var tab = tabutil.NewReader("VOLUME NAME\tDIRECTORY")
+		var err = tab.ParseHeader(lines[0])
+		if err != nil {
+			return err
+		}
+		for _, line := range lines {
+			name, _ := tab.ReadRow(line, "VOLUME NAME")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_ls_test.go:124]: Probably you should use `--quiet` for simplifying the tests

**Comment by AkihiroSuda** [comment]
Please squash commits

File: README.md (hunk 1 above)
Comment (Zheaoli) [inline README.md:1087]: maybe we can and a group of operator in argument

1. size>=:int
2. size>int
3. size<=int
4. size<int

File: cmd/nerdctl/volume_ls_test.go [hunk 5]
```diff
@@ -81,3 +81,276 @@ func TestVolumeLs(t *testing.T) {
...
+			}
+			numMatches++
+		}
+		if len(volNames) != numMatches {
+			return fmt.Errorf("expected %d volumes, got: %d", len(volNames), numMatches)
+		}
+		return nil
+	})
+
+	base.Cmd("volume", "ls", "--filter", "label="+tID).AssertOutWithFunc(func(stdout string) error {
```
Comment (Zheaoli) [inline cmd/nerdctl/volume_ls_test.go:140]: Just lime @AkihiroSuda use `--quiet`  is a good idea

File: README.md (hunk 1 above)
Comment (yuchanns) [reply README.md:1087]: More description plz. Don't understand.

File: cmd/nerdctl/volume_ls.go [hunk 6]
```diff
@@ -154,3 +164,54 @@ func getVolumes(cmd *cobra.Command) (map[string]native.Volume, error) {
 	}
 	return volStore.List(volumeSize)
 }
+
+func getFilters(filters []string) ([]map[string]string, []string, bool) {
+	isFilter := len(filters) > 0
+	labelFilters := make([]map[string]string, 0)
```
Comment (Zheaoli) [inline cmd/nerdctl/volume_ls.go:170]: Maybe making this as a function should be a good idea

So you can simply use it in here https://github.com/containerd/nerdctl/pull/1331/files#diff-5cd0eb713e85375bd7c5ec6da2b72eb42ebaafe5c03f582ffbbf77ee60a02b26R200-R208 like code below


```go
	for _, filterLabelFunc := range filterLabelFuncs {
                filterLabelFunc(v.Labels)
	}
```

File: README.md (hunk 1 above)
Comment (Zheaoli) [reply README.md:1087]: In volumes data structure 

```go
type Volume struct {
	Name       string             `json:"Name"`
	Mountpoint string             `json:"Mountpoint"`
	Labels     *map[string]string `json:"Labels,omitempty"`
	Size       int64              `json:"Size,omitempty"`
}
```

We record the data size for the volume. So maybe it's better to allow people filtering the volume by data size range

File: cmd/nerdctl/volume_ls_test.go [hunk 7]
```diff
@@ -81,3 +81,279 @@ func TestVolumeLs(t *testing.T) {
 	})
 
 }
+
+func TestVolumeLsFilter(t *testing.T) {
+	t.Parallel()
+	base := testutil.NewBase(t)
+	tID := testutil.Identifier(t)
+	testutil.DockerIncompatible(t)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_ls_test.go:89]: ?

Comment (yuchanns) [reply cmd/nerdctl/volume_ls_test.go]: Oops! My mistake.

File: cmd/nerdctl/volume_ls.go [hunk 8]
```diff
@@ -73,6 +76,23 @@ func volumeLsAction(cmd *cobra.Command, args []string) error {
...
+	filters, err := cmd.Flags().GetStringSlice("filter")
+	if err != nil {
+		return err
+	}
+	labelFilterFuncs, nameFilterFuncs, sizeFilterFuncs, isFilter, err := getVolumeFilterFuncs(filters)
+	if err != nil {
+		return err
+	}
+	if len(sizeFilterFuncs) > 0 && quiet {
+		logrus.Warn("cannot use --filter=size and --quiet together, ignoring --filter=size")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_ls.go:88]: Why cannot?

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: cmd/nerdctl/volume_ls.go (hunk 8 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/volume_ls.go:88]: Merging this PR, but it would be nice if we can relax this restriction

Comment (yuchanns) [reply cmd/nerdctl/volume_ls.go:88]: https://github.com/containerd/nerdctl/blob/49885ac2aad077f7e1fd10899e3b8b647640fa0c/cmd/nerdctl/volume_ls.go#L75-L78
According to previous behavior, if users enable `--quiet`, the `--size` will be disabled. And the Size field of the Volume struct is left to 0. In that case, `--filter=size` matches no volume. So I restrict those two args not to be together.

**Pull request merged**